### ISSUE TITLE
Rule idea: `jsx-unique-key`

### ISSUE THREAD

**Issue by silverwind** [yannickcr/eslint-plugin-react#2614]
I'd like to propose a rule that triggers when duplicate `key` properties are defined in on the same JSX nesting level or in arrays containing JSX at the top-level element.

React depends on unique keys and will render very unexpected results when duplicate keys are present which shoult to my knowledge almost never be intentional.

#### Examples where the rule would trigger an error:

```jsx
const div = (
  <div>
    <span key="notunique"/>
    <span key="notunique"/>
  </div>
);
```
``` jsx
const spans = [
  <span key="notunique"/>,
  <span key="notunique"/>,
];

**Comment by ljharb** [comment]
The first example shouldn't warn because `key` is irrelevant there, because it's not inside an array.

**Comment by ljharb** [comment]
I'm going to solve this by adding a `warnOnDuplicates` option.

**Comment by silverwind** [comment]
> The first example shouldn't warn because key is irrelevant there, because it's not inside an array.

I think any HTML sibling nodes can be affected by `key` conflicts.

**Comment by ljharb** [comment]
@silverwind can you provide a react codesandbox that illustrates such a problem? if so, I can try to include it in the warnOnDuplicates option prior to release.

**Comment by silverwind** [comment]
I think the worst that can happen with non-array cases is unnecessary re-renders. That said, React does warn at runtime for the first example I provided above.

> Warning: Encountered two children with the same key, `notunique`. Keys should be unique so that components maintain their identity across updates. Non-unique keys may cause children to be duplicated and/or omitted — the behavior is unsupported and could change in a future version.

**Comment by ljharb** [comment]
Ok thanks, I’ll reopen this then to cover the first case with the new option.

**Comment by tido64** [comment]
@ljharb: Since https://github.com/yannickcr/eslint-plugin-react/commit/f51ec4551b3fc85d971fc4d65414684bfabe3974, `jsx-key` is now being triggered for all elements inside a `React.Fragment`. Is that intentional? Our CI failed with 7.29: https://github.com/microsoft/rnx-kit/runs/5333921004?check_suite_focus=true

The code that it trips on: https://github.com/microsoft/rnx-kit/blob/main/packages/test-app/src/App.native.tsx#L39

**Comment by Lexe003** [comment]
It triggers now on all elements inside my React app - also on elements that are not in an array.

**Comment by stefcameron** [comment]
With this, I have now 7195 new lint errors for things that are completely not related to an array that never tripped before. For example:

```
            <Fragment>
              <div data-testid={`${testId}-message`}>{error.message}</div>
              <div data-testid={`${testId}-stack`}>{error.stack}</div>
            </Fragment>
```

Causes:

```
  49:15  error  Missing "key" prop for element in array  react/jsx-key
  49:15  error  Missing "key" prop for element in array  react/jsx-key
  50:15  error  Missing "key" prop for element in array  react/jsx-key
  50:15  error  Missing "key" prop for element in array  react/jsx-key
```

**Comment by ljharb** [comment]
@stefcameron yes, thank you. that's #3215.

**Comment by ljharb** [comment]
@tido64 if you're getting a different problem than #3215, please file a new issue.
